Add type checking to `has`

noahmorrison opened the pull request:
$has now throws an error when a key is not a string or an integer

Closes issue #68 on github

Here's an example of new behavior. 4 still matches "4". Don't know if that should be fixed?

Using haskell syntax highlighting because I don't know of anything closer

``` haskell
let a = global has "console"
let b = {"4": "four"} has "4"
let c = {"4": "four"} has 4
let d = {"4": "four"} has 4.0

#-- These both throw runtime errors
#-- let e = {"4": "four"} has 4.1
#-- let f = global has []
let results = a.toString() ++ b.toString() ++ c.toString() ++ d.toString()
in global.console.log(results)
```

noahmorrison committed:
Add type checking to `has`

$has now throws an error when a key is not a string or an integer

Closes issue #68 on github

wavebeem commented:
I think `{"4": true} has 4` being true is perfectly acceptable. Squiggle could do something silly like only allow integer keys if it's working with an array, but I think that's a bad idea given how many JS things return array-like objects, and it just makes things more complicated.

I think maybe the logic could be simplified into a single block to accept proper values, something like `typeof key === "string" || (typeof key === "number" && key % 1 === 0)`, falling back to an error like `"object keys must be strings or integers, got " + key`.

P.S. No worries on the syntax highlighting. Not sure how to get GitHub to support a new highlighter, hahah.

noahmorrison committed:
Simplify `has` type-guarding logic

Suggested by brian, simpler boolean logic over more precise error
messages. Which is still understandable so it works fine.

wavebeem commented:
Looks like the TravisCI build failed due to a line over 80 characters, too.

noahmorrison committed:
Reduce line length to please TravisCI

noahmorrison commented:
Made the simplification in the boolean logic. And also fixed the line length issue that my fix caused. I should really setup vim to show a line at 80 characters :stuck_out_tongue:

wavebeem commented:
Looks good to go. Thanks for your help! :+1:
